### PR TITLE
Signed Message Implementation

### PR DESCRIPTION
**Pull request by asmarty**

### CODE REVIEW

File: src/main/kotlin/org/dif/DIDComm.kt [hunk 1]
```diff
@@ -133,9 +146,27 @@ class DIDComm(private val didDocResolver: DIDDocResolver, private val secretReso
...
-        )
+        val didDocResolver = params.didDocResolver ?: this.didDocResolver
+        val secretResolver = params.secretResolver ?: this.secretResolver
+        val recipientKeySelector = RecipientKeySelector(didDocResolver, secretResolver)
+
+        val parseResult = parse(params.packedMessage)
+
+        return when (parseResult.typ) {
+            Typ.Signed -> let {
+                val kid = parseResult.jws.unprotectedHeader.keyID
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/DIDComm.kt:157]: What is `unprotectedHeader` is Null?

File: src/main/kotlin/org/dif/DIDComm.kt [hunk 2]
```diff
@@ -133,9 +146,27 @@ class DIDComm(private val didDocResolver: DIDDocResolver, private val secretReso
...
+                val key = recipientKeySelector.verifyKey(kid)
+
+                UnpackResult(
+                    verify(parseResult.jws, key),
+                    Metadata(false, false, false, false, false, listOf())
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/DIDComm.kt:162]: 1) I would specify every parameter:
```
Metadata(encrypted=false, authenticated=false, nonRepudiation=true,...)
```
2) In Metdadata: nonRepudiation=true

File: src/main/kotlin/org/dif/crypto/Parse.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+class ParseResult {
+    private var _jwm: Message? = null
+    private var _jws: JWSObjectJSON? = null
+    private var _jwe: JWEObjectJSON? = null
+
+    var typ: Typ
+        private set
+
+    val jwm: Message
+        get() = _jwm ?: throw DIDCommException("JWM should be specified")
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/crypto/Parse.kt:30]: Did you consider alternatives:
1) Algebraic Data Types (for example https://medium.com/sharenowtech/kotlin-adt-74472319962a)
2) Just separate classes for `ParseResultJWM`, `ParseResultJWE`, `ParseResultPlain`, and something like 
```
when (parseResult) {
    is ParseResultJWM -> ...
    is ParseResultJWE -> ...
    is ParseResultPlain -> ...
}
```

File: src/main/kotlin/org/dif/crypto/Verify.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import org.dif.exceptions.UnsupportedAlgorithm
+import org.dif.exceptions.UnsupportedCurveException
+import org.dif.exceptions.UnsupportedJWKException
+import org.dif.message.Message
+import org.dif.utils.asKey
+
+fun verify(jws: JWSObjectJSON, key: Key): Message {
+    val jwk = key.toJWK()
+
+    val alg = when (jwk) {
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/crypto/Verify.kt:21]: Looks like a common part between `sign` and `verify`.

File: src/main/kotlin/org/dif/exceptions/Exceptions.kt [hunk 5]
```diff
@@ -15,9 +15,60 @@ open class DIDCommException(message: String, cause: Throwable? = null) : Throwab
...
+ */
+open class DIDDocException(message: String) : DIDCommException(message)
+
 /**
  * This exception SHOULD be raised if DID Doc can not be found.
  *
  * @param did The did.
  */
-class DIDDocNotFoundException(did: String) : DIDCommException("The DID Doc $did not found")
+class DIDDocNotFoundException(did: String) : DIDDocException("The DID Doc $did is not found")
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/exceptions/Exceptions.kt:30]: I believe we decided to call it `DIDDocNotResolvedException`

File: src/main/kotlin/org/dif/crypto/key/SenderKeySelector.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+class SenderKeySelector(private val didDocResolver: DIDDocResolver, private val secretResolver: SecretResolver) {
+    fun signKey(selector: String): Key = Key.wrapSecret(
+        if (isDIDFragment(selector)) {
+            secretResolver.findKey(selector).orElseThrow { throw SecretNotFoundException(selector) }
+        } else {
+            val didDoc = didDocResolver.resolve(selector).orElseThrow { throw DIDDocNotFoundException(selector) }
+
+            val authentication = didDoc.authentications.firstOrNull()
+                ?: throw DIDDocException("Authentication is not found in DID Doc '$selector'")
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/crypto/key/SenderKeySelector.kt:18]: Should we raise `DIDUrlNotFoundException`?

File: src/main/kotlin/org/dif/crypto/key/RecipientKeySelector.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+            val (did) = divideDIDFragment(selector)
+            val didDoc = didDocResolver.resolve(did).orElseThrow { throw DIDDocNotFoundException(did) }
+
+            val verificationMethodId = if (isDIDFragment(selector)) { selector } else {
+                didDoc.authentications.firstOrNull()
+                    ?: throw DIDDocException("Authentication is not found in DID Doc '$did'")
+            }
+
+            val verificationMethod = didDoc.verificationMethods.find { it.id == verificationMethodId }
+                ?: throw DIDDocException("Verification method '$verificationMethodId' is not found in DID Doc '$did'")
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/crypto/key/RecipientKeySelector.kt:23]: Should we raise `DIDUrlNotFoundException`?

File: src/main/kotlin/org/dif/exceptions/Exceptions.kt [hunk 8]
```diff
@@ -15,9 +15,60 @@ open class DIDCommException(message: String, cause: Throwable? = null) : Throwab
...
+ */
+class UnsupportedAlgorithm(alg: String) : DIDCommException("The algorithm $alg is unsupported")
+
+/**
+ * Signals that an error has been reached unexpectedly while parsing.
+ *
+ * @param message - the detail message.
+ * @param cause - the cause of this.
+ */
+class ParseException(message: String, cause: Throwable) : DIDCommException(message, cause)
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/exceptions/Exceptions.kt:67]: Can we use `MalformedMessageException` instead?

File: src/main/kotlin/org/dif/exceptions/Exceptions.kt [hunk 9]
```diff
@@ -15,9 +15,60 @@ open class DIDCommException(message: String, cause: Throwable? = null) : Throwab
...
+ * @param kid The Key Identifier.
+ */
+class SecretNotFoundException(kid: String) : DIDCommException("The Secret Doc $kid is not found")
+
+/**
+ * This exception SHOULD be raised if curve is not supported.
+ *
+ * @param curve The curve.
+ */
+class UnsupportedCurveException(curve: String) : DIDCommException("The curve $curve is unsupported")
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/exceptions/Exceptions.kt:44]: Not sure about it, but is it better to have just one `UnsupportedXXX` exception?

File: src/main/kotlin/org/dif/message/Message.kt [hunk 10]
```diff
@@ -39,6 +41,17 @@ data class Message(
...
+
+        fun parse(json: Map<String, Any>): Message = builder(
+            JSONObjectUtils.getString(json, "id"),
+            JSONObjectUtils.getJSONObject(json, "body"),
+            JSONObjectUtils.getString(json, "type")
+        )
+            .from(JSONObjectUtils.getString(json, "from"))
+            .to(JSONObjectUtils.getStringList(json, "to"))
+            .createdTime(JSONObjectUtils.getInt(json, "created_time"))
+            .expiresTime(JSONObjectUtils.getInt(json, "expires_time"))
```
Comment (ashcherbakov) [inline src/main/kotlin/org/dif/message/Message.kt:53]: What about other fields? Is it TBD?

File: src/test/kotlin/org/dif/SignedMessageTest.kt [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            .from(ALICE_DID)
+            .to(listOf(BOB_DID))
+            .createdTime(1516269022)
+            .expiresTime(1516385931)
+            .build()
+
+        val TEST_VECTORS = listOf(
+            TestVector(
+                from = "did:example:alice#key-1",
+                expected = """
```
Comment (ashcherbakov) [inline src/test/kotlin/org/dif/SignedMessageTest.kt:34]: Did you consider putting test vectors into JSON files? We can just load it from there and trim whitespaces.

File: src/test/kotlin/org/dif/SignedMessageTest.kt [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+    companion object {
+        const val ALICE_DID = "did:example:alice"
+        const val BOB_DID = "did:example:bob"
+
+        private const val ID = "1234567890"
+        private const val TYPE = "http://example.com/protocols/lets_do_lunch/1.0/proposal"
+        private val BODY = mapOf("messagespecificattribute" to "and its value")
+
+        val PLAINTEXT_MESSAGE = Message.builder(ID, BODY, TYPE)
```
Comment (ashcherbakov) [inline src/test/kotlin/org/dif/SignedMessageTest.kt:24]: Not critical now, rather for future: I believe we will re-use the Plaintext in all tests against test vectors, so may make sense to move it into a common place.

File: src/test/kotlin/org/dif/SignedMessageTest.kt [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+            .from(ALICE_DID)
+            .to(listOf(BOB_DID))
+            .createdTime(1516269022)
+            .expiresTime(1516385931)
+            .build()
+
+        val TEST_VECTORS = listOf(
```
Comment (ashcherbakov) [inline src/test/kotlin/org/dif/SignedMessageTest.kt:31]: Is it test vectors from our DIDComm examples?

File: src/test/kotlin/org/dif/SignedMessageTest.kt [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+                          }
+                       ]
+                    }
+                """.trimIndent()
+            )
+        )
+    }
+
+    @Test
+    fun test_signed_message_test_vectors() {
```
Comment (ashcherbakov) [inline src/test/kotlin/org/dif/SignedMessageTest.kt:89]: Are you planning to add negative tests here? (kid not found, malformed message, did not found, signature is invalid, etc.)

File: src/test/kotlin/org/dif/SignedMessageTest.kt (hunk 13 above)
Comment (asmarty) [reply src/test/kotlin/org/dif/SignedMessageTest.kt:31]: yes

File: src/main/kotlin/org/dif/crypto/Sign.kt [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        JWSAlgorithm.ES256K -> ECDSASigner(jwk.asKey<ECKey>())
+        JWSAlgorithm.EdDSA -> Ed25519Signer(jwk.asKey())
+        else -> throw UnsupportedAlgorithm(alg.name)
+    }
+
+    val jwsHeader = JWSHeader.Builder(alg)
+        .type(JOSEObjectType(Typ.Signed.typ))
+        .build()
+
+    val jws = JWSObjectJSON(jwsHeader, Payload(payload))
```
Comment (CluEleSsUK) [inline src/main/kotlin/org/dif/crypto/Sign.kt:46]:
```
return JWSObjectJSON(jwsHeader, Payload(payload))
    .apply { sign(UnprotectedHeader.Builder(key.id).build(), signer) }
    .serialize()
```
would get rid of an extra `val` and make it clear the sign is  a side-effect

File: src/main/kotlin/org/dif/diddoc/DIDDocResolverInMemory.kt [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package org.dif.diddoc
+
+import java.util.Optional
+
+class DIDDocResolverInMemory(private val docs: Map<String, DIDDoc>) : DIDDocResolver {
+    constructor(docs: List<DIDDoc>) : this(docs.map { it.did to it }.toMap())
+
+    override fun resolve(did: String): Optional<DIDDoc> =
```
Comment (CluEleSsUK) [inline src/main/kotlin/org/dif/diddoc/DIDDocResolverInMemory.kt:8]: `Optional.fromNullable(docs[did])`

File: src/main/kotlin/org/dif/secret/SecretResolverInMemory.kt [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.Optional
+
+class SecretResolverInMemory(private val secrets: Map<String, Secret>) : SecretResolver {
+    constructor(docs: List<Secret>) : this(docs.map { it.kid to it }.toMap())
+
+    override fun findKey(kid: String): Optional<Secret> =
+        secrets[kid]?.let { Optional.of(it) } ?: Optional.empty()
+
+    override fun findKeys(kids: List<String>): List<Secret> =
+        kids.map { findKey(it) }.filter { it.isPresent }.map { it.get() }
```
Comment (CluEleSsUK) [inline src/main/kotlin/org/dif/secret/SecretResolverInMemory.kt:12]: `kids.mapNotNull { findKey(it).orNull() }` is a wee bit shorter

File: src/main/kotlin/org/dif/secret/SecretResolverInMemory.kt [hunk 18]
```diff
@@ -0,0 +1,13 @@
+package org.dif.secret
+
+import java.util.Optional
+
+class SecretResolverInMemory(private val secrets: Map<String, Secret>) : SecretResolver {
+    constructor(docs: List<Secret>) : this(docs.map { it.kid to it }.toMap())
+
+    override fun findKey(kid: String): Optional<Secret> =
+        secrets[kid]?.let { Optional.of(it) } ?: Optional.empty()
```
Comment (CluEleSsUK) [inline src/main/kotlin/org/dif/secret/SecretResolverInMemory.kt:9]: `Optional.fromNullable(secrets[kid])`

File: src/test/kotlin/org/dif/DIDCommTest.kt [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+        val PLAINTEXT_MESSAGE = Message.builder(ID, BODY, TYPE)
+            .from(ALICE_DID)
+            .to(listOf(BOB_DID))
+            .createdTime(1516269022)
+            .expiresTime(1516385931)
+            .build()
+    }
+
+    @Test
+    fun test_message_with_attachments() {
```
Comment (CluEleSsUK) [inline src/test/kotlin/org/dif/DIDCommTest.kt:34]: you can write nice test names inside ``, e.g.

```
fun `this is a cool readable test name`() {
  //
}
```

**Review by CluEleSsUK — APPROVED** [review]
Some small stylistic suggestions, but otherwise lgtm!

File: src/main/kotlin/org/dif/message/Message.kt (hunk 10 above)
Comment (asmarty) [reply src/main/kotlin/org/dif/message/Message.kt:53]: yes, I suggest implementing it later

File: src/test/kotlin/org/dif/SignedMessageTest.kt (hunk 14 above)
Comment (asmarty) [reply src/test/kotlin/org/dif/SignedMessageTest.kt:89]: I suggest adding negative tests later

File: src/main/kotlin/org/dif/crypto/key/SenderKeySelector.kt (hunk 6 above)
Comment (asmarty) [reply src/main/kotlin/org/dif/crypto/key/SenderKeySelector.kt:18]: I will change Exceptions later

File: src/main/kotlin/org/dif/crypto/key/RecipientKeySelector.kt (hunk 7 above)
Comment (asmarty) [reply src/main/kotlin/org/dif/crypto/key/RecipientKeySelector.kt]: I will change Exceptions later

File: src/main/kotlin/org/dif/exceptions/Exceptions.kt (hunk 8 above)
Comment (asmarty) [reply src/main/kotlin/org/dif/exceptions/Exceptions.kt:67]: I will change Exceptions later, But in my opinion it's an error that was happened during parsing message

File: src/main/kotlin/org/dif/exceptions/Exceptions.kt (hunk 9 above)
Comment (asmarty) [reply src/main/kotlin/org/dif/exceptions/Exceptions.kt:44]: I will change Exceptions later

File: src/test/kotlin/org/dif/SignedMessageTest.kt (hunk 11 above)
Comment (asmarty) [reply src/test/kotlin/org/dif/SignedMessageTest.kt]: I created a fixtures package

**Review by ashcherbakov — APPROVED** [review]

**Pull request merged**